src/Databases/AlterDatabaseCommands.cpp

```diff
@@ -0,0 +1,413 @@
+            table_override->table_name);
+    }
+    apply_data.overrides_already_seen_in_query.insert(table_override->table_name);
+    ASTTableOverride * existing_override = tryGetExistingOverride(apply_data.getCreateQuery(), table_override->table_name);
+    if ((type == ASTAlterCommand::DROP_TABLE_OVERRIDE || type == ASTAlterCommand::MODIFY_TABLE_OVERRIDE) && !existing_override)
+    {
+        throw Exception(
+            ErrorCodes::TABLE_OVERRIDE_NOT_FOUND,
+            "ALTER DATABASE {} TABLE OVERRIDE refers to unknown override `{}`",
+            whichOverrideCommand(type),
+            table_override->table_name);
+    }
+    else if (type == ASTAlterCommand::ADD_TABLE_OVERRIDE && existing_override)
+    {
+        throw Exception(
+            ErrorCodes::TABLE_OVERRIDE_ALREADY_EXISTS,
+            "ALTER DATABASE {} TABLE OVERRIDE refers to existing override `{}`",
+            whichOverrideCommand(type),
+            table_override->table_name);
+    }
+
+    if (database->isTableExist(table_override->table_name, context))
+    {
+        if (table_override->modifiesStorage() != (existing_override && existing_override->modifiesStorage()))
+        {
+            /// Only the old or new override modifies storage
+            apply_data.need_to_recreate_database = true;
+            if (table_override->modifiesStorage())
+                apply_data.need_to_recreate_database_reason = "existing override does not modify storage, new override does";
+            else
+                apply_data.need_to_recreate_database_reason = "existing override modifies storage, new override does not";
+
+        }
+        else if (table_override->modifiesStorage())
+        {
+            /// Here, we know existing_override->storage exists after the previous if. If storage parameters
+            /// are different, the database needs to be recreated.
+            if (serializeAST(*table_override->storage) != serializeAST(*existing_override->storage))
+            {
+                apply_data.need_to_recreate_database = true;
+                apply_data.need_to_recreate_database_reason = "new override modifies storage";
+            }
+        }
+    }
+    /// TODO: check that no overrides touch internal columns like _version, _snapshot[12] and _sign
+    ///       (but account for that these names could be _version2 etc.)
+    /// TODO: check that no overrides touch internal indices like _version
+
+    if (auto alter_table_query = tryGetAlterTableQuery(database, apply_data, context))
+    {
+        apply_data.reconciliation_queries.push_back(alter_table_query);
+    }
+
+}
+
+void AlterDatabaseTableOverrideCommand::apply(DatabasePtr &, ContextPtr, AlterDatabaseApplyData & apply_data)
+{
+    ASTCreateQuery * create = apply_data.getCreateQuery();
+    if (type == ASTAlterCommand::ADD_TABLE_OVERRIDE || type == ASTAlterCommand::MODIFY_TABLE_OVERRIDE)
+    {
+        if (!create->table_overrides)
```

Comment:
I'd assume that if `type == ASTAlterCommand::MODIFY_TABLE_OVERRIDE` then `create->table_overrides` always `!= nullptr`, isn't it? Should this be asserted then?

Reply:
Yeah, I was just being lazy. Will use chassert().